src/struct_mv/boxloop_sycl.h

```diff
@@ -0,0 +1,429 @@
+   hypre_BasicBoxLoopDataDeclareK(1, ndim, loop_size, stride1);                                       \
+   BoxLoopforall(hypre__tot, [=] (sycl::nd_item<1> item)                                              \
+   {                                                                                                  \
+      HYPRE_Int idx = (HYPRE_Int) item.get_global_linear_id();                                        \
+      if (idx < hypre__tot)                                                                           \
+      {                                                                                               \
+         hypre_newBoxLoopDeclare(databox1);                                                           \
+         hypre_BoxLoopIncK(1, databox1, i1);
+
+/* BoxLoop 2 */
+#define hypre_newBasicBoxLoop2Begin(ndim, loop_size, stride1, i1, stride2, i2)                        \
+{                                                                                                     \
+   hypre_newBoxLoopInit(ndim, loop_size);                                                             \
+   hypre_BasicBoxLoopDataDeclareK(1, ndim, loop_size, stride1);                                       \
+   hypre_BasicBoxLoopDataDeclareK(2, ndim, loop_size, stride2);                                       \
+   BoxLoopforall(hypre__tot, [=] (sycl::nd_item<1> item)                                              \
+   {                                                                                                  \
+      HYPRE_Int idx = (HYPRE_Int) item.get_global_linear_id();                                        \
+      if (idx < hypre__tot)                                                                           \
+      {                                                                                               \
+         hypre_newBoxLoopDeclare(databox1);                                                           \
+         hypre_BoxLoopIncK(1, databox1, i1);                                                          \
+         hypre_BoxLoopIncK(2, databox2, i2);
+
+
+/* Reduction BoxLoop1 */
+#define hypre_newBoxLoop1ReductionBegin(ndim, loop_size, dbox1, start1, stride1, i1, sum_var)         \
+{                                                                                                     \
+   hypre_newBoxLoopInit(ndim, loop_size);                                                             \
+   hypre_BoxLoopDataDeclareK(1, ndim, loop_size, dbox1, start1, stride1);                             \
+   HYPRE_Real *shared_sum_var = hypre_CTAlloc(HYPRE_Real, 1, HYPRE_MEMORY_DEVICE);                    \
+   hypre_TMemcpy(shared_sum_var, &sum_var, HYPRE_Real, 1, HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_HOST);    \
+   ReductionBoxLoopforall( [=] (sycl::nd_item<1> item, auto &hypre_sycl_sum)                          \
+   {                                                                                                  \
```

Comment:
@waynemitchell is it possible to use `sum_var` here, in the place of `hypre_sycl_sum`, so the interface can be simpler, i.e., don't need the `#if defined(HYPRE_USING_SYCL)` at calling reduction? I am not very familiar with sycl yet. Just curious.

Reply:
@liruipeng I couldn't find a cleaner way of implementing this (e.g. without the `#if defined(HYPRE_USING_SYCL)` macros at the reduction calls). `hypre_sycl_sum` here is actually a `sycl::reducer` variable. Inside the lambda (i.e. in the contents of the boxloop) you need to use the variable name of the `sycl::reducer`. I could see doing something similar to the cuda implementation here, i.e. at the call, setup a reducer object to pass to the boxloop. But this just moves where you have to put the `HYPRE_USING_SYCL` macro guards at the reduction call. I will probably do something like this in the future (in particular, if doing multiple reduction simultaneously works as expected with the newer intel compilers), but I think the current implementation works and is good enough for now.

Reply:
@waynemitchell Maybe I wasn't clear about what I was asking about. Can we just write `auto &sum_var` instead of `auto &hypre_sycl_sum` here, if it makes sense at all? I **guess** `hypre_sycl_sum` is an argument (as you said, a reducer type) of the lambda and the name is up to our choice (did not find it used anywhere else). If this can work, do we still need the `#if defined` at the call? Please let me know if this makes sense. Thanks!

Reply:
@liruipeng If I understand you correctly, you're suggesting that we use `sum_var` as the name for the reducer variable. I expected this to cause conflicts since `sum_var` is also a variable name captured by the lambda, correct? I tried this, and it actually seems to work, however. I'm still not super comfortable working with all the details of lambdas... is this definitely safe?

Reply:
@liruipeng What about the first answer here:
https://stackoverflow.com/questions/58285806/how-to-pass-local-variable-and-argument-of-same-name-using-this-keyword-inside
I can do something like:
`[=,hypre_unused_var=sum_var] (sycl::nd_item<1> item, auto &sum_var)`
That way, there should be no conflict with the captured `sum_var` having the same name as the reducer argument passed to the lambda. Does that make sense or would you suggest something else?

Reply:
@waynemitchell Yes, that's what I suggested. Thank you for trying that. It is interesting to know it "works" or at least it does compile. I was imagining (ab)use `sum_var` using the same variable name at different scopes. Essentially, we have something like
```
int var;
auto glambda = [](auto var) { var += ...};
auto b = glambda(...)

```
I guess the second `var` as the function arg. takes precedence over the first one inside the lambda (?). Maybe we should confirm this with Intel. But I am also happy with the current implementation. So, I approved this PR. Thank you! This is a nice work.

Reply:
@waynemitchell Thank you for the reference. I think `[=,hypre_unused_var=sum_var] (sycl::nd_item<1> item, auto &sum_var)` should be the way to go. We will use `sum_var` safely in loop body. In this case, we don't actually have to capture the value of `sum_var` from outer scope, since I see an H2D copy is performed (is it right?). Thanks again!

Reply:
@liruipeng OK, I just pushed changes that implement this idea. This really cleans up the calls to the reduction boxloop, which I like a lot. Thanks for the suggestion!

Reply:
Superb! Thank you for doing it.